pkg/generic/thrift/json_go116plus_amd64.go

```diff
@@ -0,0 +1,152 @@
+	"github.com/bytedance/gopkg/lang/mcache"
+	"github.com/cloudwego/dynamicgo/conv/j2t"
+
+	"github.com/cloudwego/kitex/pkg/generic/descriptor"
+	"github.com/cloudwego/kitex/pkg/remote/codec/perrors"
+	cthrift "github.com/cloudwego/kitex/pkg/remote/codec/thrift"
+	"github.com/cloudwego/kitex/pkg/utils"
+)
+
+// Write write json string to out thrift.TProtocol
+func (m *WriteJSON) Write(ctx context.Context, out thrift.TProtocol, msg interface{}, requestBase *Base) error {
+	// fallback logic
+	if !m.dynamicgoEnabled {
+		return m.originalWrite(ctx, out, msg, requestBase)
+	}
+
+	var cv j2t.BinaryConv
+	// dynamicgo logic
+	if !m.hasRequestBase {
+		requestBase = nil
+	}
+	if requestBase != nil {
+		cv = m.j2tBinaryConvWithThriftBase
+	} else {
+		cv = m.j2tBinaryConv
+	}
+
+	// msg is void
+	if _, ok := msg.(descriptor.Void); ok {
+		if err := m.writeHead(out); err != nil {
+			return err
+		}
+		if err := m.writeFields(ctx, out, Void, nil, nil, nil); err != nil {
+			return err
+		}
+		return writeTail(out)
+	}
+
+	// msg is string
+	s, ok := msg.(string)
+	if !ok {
+		return perrors.NewProtocolErrorWithType(perrors.InvalidData, "decode msg failed, is not string")
+	}
+
+	transBuff := utils.StringToSliceByte(s)
+	dbuf := mcache.Malloc(len(transBuff))[0:0]
+	defer mcache.Free(dbuf)
+
+	if err := m.writeHead(out); err != nil {
+		return err
+	}
+	if err := m.writeFields(ctx, out, String, &cv, transBuff, dbuf); err != nil {
+		return err
+	}
+	return writeTail(out)
+}
+
+type MsgType int
+
+const (
+	Void MsgType = iota
+	String
+)
+
+func (m *WriteJSON) writeFields(ctx context.Context, out thrift.TProtocol, msgType MsgType, cv *j2t.BinaryConv, transBuff, dbuf []byte) error {
+	for _, field := range m.dynamicgoTypeDsc.Struct().Fields() {
+		// Exception field
+		if !m.isClient && field.ID() != 0 {
+			// generic server ignore the exception, because no description for exception
+			// generic handler just return error
+			continue
+		}
+
+		if err := out.WriteFieldBegin(field.Name(), field.Type().Type().ToThriftTType(), int16(field.ID())); err != nil {
+			return err
+		}
+		switch msgType {
+		case Void:
+			if err := out.WriteStructBegin(field.Name()); err != nil {
+				return err
+			}
+			if err := out.WriteFieldStop(); err != nil {
+				return err
+			}
+			if err := out.WriteStructEnd(); err != nil {
+				return err
+			}
+		case String:
+			// encode using dynamicgo
+			// json []byte to thrift []byte
+			if err := cv.DoInto(ctx, field.Type(), transBuff, &dbuf); err != nil {
+				return err
+			}
+		}
+		// WriteFieldEnd has no content
+		// if err := out.WriteFieldEnd(); err != nil {
+		// 	return err
+		// }
+	}
+	if msgType == String {
+		tProt, ok := out.(*cthrift.BinaryProtocol)
+		if !ok {
+			return perrors.NewProtocolErrorWithMsg("TProtocol should be BinaryProtocol")
+		}
+		buf, err := tProt.ByteBuffer().Malloc(len(dbuf))
+		if err != nil {
+			return err
+		}
+		// TODO: implement MallocAck() to achieve zero copy
+		copy(buf, dbuf)
```

Comment:
why copying `dbuf` to `buf` here? `buf` is not used anymore

Reply:
@felix021 It's actually writing into `out`. We need to write the buf with real written length into out but we don't have MallocAck yet, so I define `buf, err := tProt.ByteBuffer().Malloc(len(dbuf))` and write the buf into `out` by copying here.

Reply:
@felix021 Since `WriteBinary` might reference the original buf, we can't mcache.Free buf directly, so I use this API to replace WriteBinary so that we can mcache.Free(buf) to reduce memory alloc.

Reply:
Oh, it might be netpoll nocopy buffer which will keep reference of the bytes.

So your plan is: 

before calling `writeFields`
1) malloc `dbuf` with mcache

in `writeFields`
2) write thrift to `dbuf`
3) malloc `buf` with tProf.ByteBuffer()
4) copy from `dbuf` to `buf`

after `writeFields`
5) recycle `dbuf` with mcache

You see, we can simple do it by:

in `writeFields`
1) malloc `buf` with tProf.ByteBuffer()
2) write thrift to `buf`

Reply:
`DoInto` can modify the buf, so we need to use another buf for `DoInto` besides buf getting from `tProf.ByteBuffer.Malloc`